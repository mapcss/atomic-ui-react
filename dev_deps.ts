export {
  anyBoolean,
  anyFunction,
  anyNumber,
  anyString,
  defineExpect,
  defineGlobalThis,
  fn,
  jestMatcherMap,
  jestModifierMap,
} from "https://deno.land/x/unitest@v1.0.0-beta.82/mod.ts";
import {
  defineExpect,
  equal,
  jestMatcherMap,
  jestModifierMap,
  MatchResult,
} from "https://deno.land/x/unitest@v1.0.0-beta.82/mod.ts";
export { FakeTime } from "https://deno.land/std@0.136.0/testing/time.ts";
export { describe, it } from "https://deno.land/std@0.136.0/testing/bdd.ts";
export { assertSnapshot } from "https://deno.land/std@0.136.0/testing/snapshot.ts";
export { default as clsx } from "https://esm.sh/clsx@1.1.1?pin=v78";
export async function setupJSDOM(): Promise<void> {
  const { JSDOM } = await import("https://esm.sh/jsdom@19.0.0?pin=v76").then((
    module,
  ) => module.default);

  const doc = new JSDOM(`<!DOCTYPE html>`);
  globalThis.document = doc.window.document;
  globalThis.HTMLIFrameElement = doc.window.HTMLIFrameElement;
  globalThis.Node = doc.window.Node;
}
export const expect = defineExpect({
  matcherMap: {
    ...jestMatcherMap,
    toHaveFocus,
    toHaveAttribute,
  },
  modifierMap: jestModifierMap,
});

function toHaveFocus(value: Element): MatchResult {
  return {
    pass: value.ownerDocument.activeElement === value,
    expected: "Actual element is focused",
  };
}

function toHaveAttribute(
  el: Element,
  name: string,
  expected: unknown,
): MatchResult {
  if (!el.hasAttribute(name)) {
    return {
      pass: false,
      expected: `[${name}] is not exists`,
    };
  }
  return {
    pass: equal(el.getAttribute(name), expected),
    expected: `${name} is ${Deno.inspect(expected)}`,
  };
}

export function setupRaf(): () => void {
  Object.defineProperty(globalThis.window, "requestAnimationFrame", {
    value: (handler: TimerHandler) => globalThis.setTimeout(handler, 1),
    configurable: true,
    writable: true,
  });

  Object.defineProperty(globalThis.window, "cancelAnimationFrame", {
    value: globalThis.clearTimeout,
    configurable: true,
    writable: true,
  });

  const reset = (): void => {
    globalThis.requestAnimationFrame = _.requestAnimationFrame;
    globalThis.cancelAnimationFrame = _.cancelAnimationFrame;
  };

  return reset;
}

// deno-lint-ignore no-explicit-any
export type ParamReturn<T extends (...args: any[]) => unknown> = [
  ...Parameters<T>,
  ReturnType<T>,
];

const _: {
  getComputedStyle: typeof window.getComputedStyle;
  requestAnimationFrame: typeof globalThis.requestAnimationFrame;
  cancelAnimationFrame: typeof cancelAnimationFrame;
} = {
  getComputedStyle: globalThis.getComputedStyle,
  requestAnimationFrame: globalThis.requestAnimationFrame,
  cancelAnimationFrame: globalThis.cancelAnimationFrame,
};

export function mockGetComputedStyle(mock: () => CSSStyleDeclaration) {
  Object.defineProperty(globalThis.window, "getComputedStyle", {
    value: mock,
    writable: true,
    configurable: true,
  });

  const reset = (): void => {
    globalThis.getComputedStyle = _.getComputedStyle;
  };

  return reset;
}
