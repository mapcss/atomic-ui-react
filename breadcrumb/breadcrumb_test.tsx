import Breadcrumb from "./breadcrumb.ts";
import { assertSnapshot, describe, it, setupJSDOM } from "../dev_deps.ts";
import { render } from "@testing-library/react";

const describeTests = describe({
  name: "Breadcrumb",
  async beforeEach() {
    await setupJSDOM();
  },
});

it(describeTests, "should be match snapshot", (t) => {
  const { container } = render(
    <Breadcrumb>
      <a>test</a>
    </Breadcrumb>,
  );

  assertSnapshot(t, container.innerHTML);
});

it(describeTests, "should add aria-current attribute to last element", (t) => {
  const { container } = render(
    <Breadcrumb>
      <a>1</a>
      <a>2</a>
      <a>3</a>
    </Breadcrumb>,
  );

  assertSnapshot(t, container.innerHTML);
});

it(describeTests, "should change separator", (t) => {
  const { container } = render(
    <Breadcrumb separator={<span>{">"}</span>}>
      <a>1</a>
      <a>2</a>
    </Breadcrumb>,
  );

  assertSnapshot(t, container.innerHTML);
});

it(describeTests, "should overwrite components", (t) => {
  const { container, rerender } = render(
    <Breadcrumb
      components={{
        nav: ({ children }) => <nav children={children} />,
      }}
    >
      <a>1</a>
      <a>2</a>
    </Breadcrumb>,
  );

  assertSnapshot(t, container.innerHTML);

  rerender(
    <Breadcrumb
      components={{
        ol: (props) => <ul {...props} />,
      }}
    >
      <a>1</a>
      <a>2</a>
    </Breadcrumb>,
  );

  assertSnapshot(t, container.innerHTML);

  rerender(
    <Breadcrumb
      components={{
        li: (props, { forSeparator }) =>
          forSeparator ? <></> : <li {...props} />,
      }}
    >
      <a>1</a>
      <a>2</a>
    </Breadcrumb>,
  );

  assertSnapshot(t, container.innerHTML);
});

it(describeTests, "should disabled auto adding aria current attribute", (t) => {
  const { container } = render(
    <Breadcrumb
      disabledAriaCurrent
    >
      <a>1</a>
      <a>2</a>
    </Breadcrumb>,
  );

  assertSnapshot(t, container.innerHTML);
});

it(
  describeTests,
  "should not add aria-current when the last element is not anchor",
  (t) => {
    const { container } = render(
      <Breadcrumb>
        <div>not anchor</div>
      </Breadcrumb>,
    );

    assertSnapshot(t, container.innerHTML);
  },
);
