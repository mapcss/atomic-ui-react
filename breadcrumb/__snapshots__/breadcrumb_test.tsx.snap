export const snapshot = {};

snapshot[`Breadcrumb > should be match snapshot 1`] = `'<nav aria-label="Breadcrumb"><ol><li><a href="/html">HTML</a></li><li aria-hidden="true">/</li><li><a href="/html/breadcrumb">Breadcrumb</a></li><li aria-hidden="true">/</li><li><a href="/html/breadcrumb/example" aria-current="page">Example</a></li></ol></nav>'`;

snapshot[`Breadcrumb > should add aria-current attribute to last element 1`] = `'<nav aria-label="Breadcrumb"><ol><li><a>1</a></li><li aria-hidden="true">/</li><li><a>2</a></li><li aria-hidden="true">/</li><li><a aria-current="page">3</a></li></ol></nav>'`;

snapshot[`Breadcrumb > should change separator 1`] = `'<nav aria-label="Breadcrumb"><ol><li><a>1</a></li><li aria-hidden="true"><span>&gt;</span></li><li><a aria-current="page">2</a></li></ol></nav>'`;

snapshot[`Breadcrumb > should overwrite components 1`] = `'<nav><ol><li><a>1</a></li><li aria-hidden="true">/</li><li><a aria-current="page">2</a></li></ol></nav>'`;

snapshot[`Breadcrumb > should overwrite components 2`] = `'<nav aria-label="Breadcrumb"><ul class="text-lg"><li><a>1</a></li><li aria-hidden="true">/</li><li><a aria-current="page">2</a></li></ul></nav>'`;

snapshot[`Breadcrumb > should overwrite components 3`] = `'<nav aria-label="Breadcrumb"><ol><li><a>1</a></li><li><a aria-current="page">2</a></li></ol></nav>'`;

snapshot[`Breadcrumb > should disabled auto adding aria current attribute 1`] = `'<nav aria-label="Breadcrumb"><ol><li><a>1</a></li><li aria-hidden="true">/</li><li><a>2</a></li></ol></nav>'`;

snapshot[`Breadcrumb > should not add aria-current when the last element is not anchor 1`] = `'<nav aria-label="Breadcrumb"><ol><li><div>not anchor</div></li></ol></nav>'`;
